### PR TITLE
#7841 SchemaTool generates extra diff for platforms without FK support

### PR DESCRIPTION
**Pull request by vpArth**
Add breaking test for #7841

### CODE REVIEW

**Comment by vpArth** [comment]
> `assertEmpty()`

ok, changed

**Comment by vpArth** [comment]
Commits are squashed after both(scrutinizer, travis) checks passed

File: tests/Doctrine/Tests/ORM/Functional/SchemaTool/SqliteSchemaToolTest.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional\SchemaTool;
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/SchemaTool/SqliteSchemaToolTest.php]: Please move this to the `Doctrine\Tests\ORM\Functional\Ticket` namespace

File: tests/Doctrine/Tests/ORM/Functional/SchemaTool/SqliteSchemaToolTest.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\ORM\Annotation as ORM;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Tools\SchemaTool;
+use Doctrine\Tests\OrmTestCase;
+use Doctrine\Tests\TestUtil;
+
+class SqliteSchemaToolTest extends OrmTestCase
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/SchemaTool/SqliteSchemaToolTest.php]: Please name this test class, and all related classes, after the issue it fixes (`GH7841Test`, `GH7841Parent`, `GH7841Child`) and add a `@group GH7841` annotation to the test class.

Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/SchemaTool/SqliteSchemaToolTest.php]: I think that using the `OrmFunctionalTestCase` as parent class would simplify this class a bit (remove the need for the properties and the setup method)

File: lib/Doctrine/ORM/Tools/SchemaTool.php [hunk 3]
```diff
@@ -737,33 +737,35 @@ private function gatherRelationJoinColumns(
...
+                    'foreignColumns' => $foreignColumns,
+                ];
+
+                $theJoinTable->addForeignKeyConstraint(
+                    $foreignTableName,
+                    $localColumns,
+                    $foreignColumns,
+                    $fkOptions
+                );
+            }
```
Comment (lcobucci) [inline lib/Doctrine/ORM/Tools/SchemaTool.php]: This portion of the code was already a bit complex and adding an if here just increases the cyclomatic complexity. Would you please try to use some early returns to make this a bit more readable and simpler? Like:

```php
if (! $this->platform->supportsForeignKeyConstraints()) {
    return;
}

if (isset($addedFks[$compositeName]) && ($foreignTableName !== $addedFks[$compositeName]['foreignTableName'] || array_diff($foreignColumns, $addedFks[$compositeName]['foreignColumns']) !== [])) { // expression simplified as well
    // (...)
}

if (isset($blacklistedFks[$compositeName])) {
    return;
}

$addedFks[$compositeName] = ['foreignTableName' => $foreignTableName, 'foreignColumns' => $foreignColumns];
$theJoinTable->addForeignKeyConstraint($foreignTableName, $localColumns, $foreignColumns, $fkOptions);
```

**Review by lcobucci — CHANGES_REQUESTED** [review]
@vpArth thanks for sending this patch! Here are a few comments :+1:

I also believe that this was intended as a bug fix, right? If so, it must use `2.6` as base instead of `master.

File: tests/Doctrine/Tests/ORM/Functional/SchemaTool/SqliteSchemaToolTest.php (hunk 2 above)
Comment (vpArth) [reply tests/Doctrine/Tests/ORM/Functional/SchemaTool/SqliteSchemaToolTest.php]: I get a bunch of errors in other functional tests when I use OrmFunctionalTest.  
Looks like they are used some dirty global state... 
It was only a reason to use OrmTestCase.
I'll try, ok.

File: lib/Doctrine/ORM/Tools/SchemaTool.php (hunk 3 above)
Comment (vpArth) [reply lib/Doctrine/ORM/Tools/SchemaTool.php]: I was believe, it should be in refactoring stage unrelated to any issue. But ok, will do it.

**Comment by vpArth** [comment]
> it must use `2.6` as base instead of `master.

vendor/bin/phpcs fails every single file on 2.6 branch. is it ok?

**Comment by lcobucci** [comment]
> > it must use `2.6` as base instead of `master.
> 
> vendor/bin/phpcs fails every single file on 2.6 branch. is it ok?

@vpArth we have configured it to validate the diff in PRs because of that. You can focus on your changes only and ensure they're correct.

**Comment by vpArth** [comment]
Should I migrate to new PR to switch target branch to 2.6?

**Review by lcobucci — APPROVED** [review]
Awesome :tada:
Thanks!

**Comment by beberlei** [comment]
This is causing a regression in 2.6.5 and 2.7 for SQLite users:

**Comment by beberlei** [comment]
This needs to be reverted. It fixes the symptom and not the problem.

1. SQLite does not support foreign keys, but we emulate them. A check for `supportsForeignKeyConstraints()` is the wrong place, because we interact with DBAL SchemaManager API in SchemaTool that abstracts the platform and if foriegn keys are emulated or supported.

2. Debugging into the code, the culprit is `Comparator::diffTable()` detects an "added foreign key", even though it already exists. That means that the SQLite abstraction does not populate `Table#foreignKeys` with the emulated keys.

3. The real fix is probably that we let `SQLitePlatform#supportsForeignKeyConstraints()` return true.

**Comment by beberlei** [comment]
Fix in DBAL: https://github.com/doctrine/dbal/pull/3762

**Comment by lcobucci** [comment]
@vpArth we had to revert the changes as @beberlei explained. Unfortunately the fix can only be applied for ORM v2.8.0 since it requires us to bump up DBAL version (to 2.10).

**Comment by vpArth** [comment]
But what about #7841? 
The SchemaTool bug for all platforms with false in `supportsForeignKeyConstraints` is still here then.

Adding true into SqlitePlatform could fix the regression without reverting that pr, right?

**Comment by beberlei** [comment]
@vpArth please see doctrine/dbal#3762 - it ads the missing support in SQLite platform when it gets merged.

**Pull request merged**